File: backend/app/schemas/clinical_copilot.py
"""Meto Clinical Copilot — doctor-facing AI decision-support schemas.

Contract-fixed response/request shapes for the four Clinical Copilot endpoints:
- ai-summary   (deterministic, no LLM)
- ai-analysis  (deterministic priority + LLM-phrased reasoning)
- ai-questions (LLM-suggested history-taking questions)
- ai-advice    (LLM-suggested counseling direction)

Every output type carries `sources` (or is itself a citation of deterministic
findings) + a `confidence` level + the mandatory disclaimer — this surface is
decision-SUPPORT only, never decision-making, and never diagnoses/prescribes.

`missing_data` + `confidence_note_vi` are populated on every response by a single
shared, deterministic completeness assessor (see
``app.services.clinical_copilot._assess_completeness``) — never by the LLM.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.domain.policies import DISCLAIMER_VI

SourceRefType = Literal[
    "lab", "metric", "medication", "condition", "allergy", "appointment", "consultation", "profile"
]
ConfidenceLevel = Literal["high", "medium", "low"]
RiskLevel = Literal["normal", "monitor", "see_doctor_soon", "urgent"]

MissingDataCategory = Literal[
    "demographics",
    "symptoms",
    "allergies",
    "medications",
    "medical_history",
    "vitals_metrics",
    "labs",
    "consultation_context",
]


class MissingDataItem(BaseModel):
    category: MissingDataCategory
    label_vi: str  # controlled label from a fixed lookup table — NEVER LLM-authored text


class SourceRef(BaseModel):
    id: str  # stable citation id. Two distinct id families:
    # - ROW-BACKED (stable/immutable for the life of the row — a fresh reading
    #   never overwrites or collides with an earlier one's id): e.g.
    #   "lab:<lab_result_id>", "metric:<health_metric_id>",
    #   "medication:<medication_id>", "consultation:<consultation_id>",
    #   "profile:<patient_id>". A "lab:<metric_type>"/"metric:<metric_type>"
    #   fallback (type-scoped, dateless) is used ONLY when no concrete row
    #   could be resolved for that finding.
    # - COLLECTION-ITEM REFERENCE (NOT stable/immutable across an edit): e.g.
    #   "condition:<index>", "allergy:<index>" — these index into a single
    #   encrypted JSON column on PatientProfile, not first-class DB rows, so
    #   the id can shift when the list is reordered/edited. Documented
    #   limitation, not a bug.
    type: SourceRefType
    label: str
    date: str | None = None  # ISO-8601; null ONLY if the underlying record genuinely
    # has no date — never invented


class CitedClaim(BaseModel):
    text: str
    sources: list[SourceRef] = Field(default_factory=list)
    basis: Literal["sourced", "needs_confirmation"]
    confidence: ConfidenceLevel  # per-claim, computed server-side from validation
    # outcome — NEVER trusted from the LLM


class MedicationBrief(BaseModel):
    name: str
    dosage: str
    frequency: str


class AbnormalFindingBrief(BaseModel):
    metric_type: str
    label: str
    status: str  # "low"|"high"|"critical"
    value_display: str
    trend_label: str
    priority: str  # "monitor"|"watch"|"see_doctor"


class ClinicalSummaryOut(BaseModel):
    as_of: str  # ISO8601
    conditions: list[str] = Field(default_factory=list)
    allergies: list[str] = Field(default_factory=list)
    medications: list[MedicationBrief] = Field(default_factory=list)
    abnormal_findings: list[AbnormalFindingBrief] = Field(default_factory=list)
    notable_changes: list[str] = Field(default_factory=list)
    sources: list[SourceRef] = Field(default_factory=list)
    confidence: ConfidenceLevel
    missing_data: list[MissingDataItem] = Field(default_factory=list)
    confidence_note_vi: str | None = None
    disclaimer: str = DISCLAIMER_VI


class RiskFlag(BaseModel):
    level: RiskLevel
    label_vi: str
    findings: list[str] = Field(default_factory=list)
    missing_data: list[MissingDataItem] = Field(default_factory=list)
    sources: list[SourceRef] = Field(default_factory=list)


class ClinicalAnalysisOut(BaseModel):
    priority: RiskFlag
    key_issues: list[CitedClaim] = Field(default_factory=list)
    contradictions_or_gaps: list[CitedClaim] = Field(default_factory=list)
    differentials_to_exclude: list[CitedClaim] = Field(default_factory=list)
    confidence: ConfidenceLevel
    missing_data: list[MissingDataItem] = Field(default_factory=list)
    confidence_note_vi: str | None = None
    disclaimer: str = DISCLAIMER_VI


QuestionGroup = Literal[
    "current_symptoms",
    "onset_timing",
    "severity_progression",
    "aggravating_relieving",
    "relevant_history",
    "medication_adherence",
    "warning_signs",
    "lifestyle",
]


class SuggestedQuestion(BaseModel):
    group: QuestionGroup
    question_vi: str
    reason_vi: str


class ClinicalQuestionsOut(BaseModel):
    questions: list[SuggestedQuestion] = Field(default_factory=list)
    confidence: ConfidenceLevel
    missing_data: list[MissingDataItem] = Field(default_factory=list)
    confidence_note_vi: str | None = None
    disclaimer: str = DISCLAIMER_VI


AdviceCategory = Literal["explain_patient", "home_monitoring", "when_to_visit", "suggested_tests"]


class AdviceItem(BaseModel):
    category: AdviceCategory
    text_vi: str


class ClinicalAdviceOut(BaseModel):
    items: list[AdviceItem] = Field(default_factory=list)
    confidence: ConfidenceLevel
    missing_data: list[MissingDataItem] = Field(default_factory=list)
    confidence_note_vi: str | None = None
    disclaimer: str = DISCLAIMER_VI


class ClinicalCopilotRequest(BaseModel):
    consultation_id: str | None = None

File: frontend/src/lib/api/clinicalCopilot.ts
import { api } from './client'

/**
 * Meto Clinical Copilot — typed client + contract.
 *
 * A doctor-facing AI decision-support panel: summarizes a patient's record
 * with sources, flags risk level, suggests history-taking questions, and
 * suggests a counseling direction. This is explicitly NOT a freeform chatbot
 * — structured cards only, never invent extra chat UI.
 *
 * PHI SAFETY (CRITICAL): every request is scoped to a single patient via
 * {@link ClinicalCopilotScope.patientId}. A request MUST NEVER carry,
 * reference, or leak data belonging to any other patient.
 *
 * When the feature flag is off (default), every getter below resolves `null`
 * with NO network request. Real `ApiError`s are never swallowed — they
 * propagate so the caller can render an error state with retry.
 */

// ---------------------------------------------------------------------------
// Shared shapes
// ---------------------------------------------------------------------------

export type SourceRef = {
  type: 'lab' | 'metric' | 'medication' | 'condition' | 'allergy' | 'appointment'
  label: string
  date?: string | null
}

export type ConfidenceLevel = 'high' | 'medium' | 'low'
export type RiskLevel = 'normal' | 'monitor' | 'see_doctor_soon' | 'urgent'

export type MedicationBrief = { name: string; dosage: string; frequency: string }
export type AbnormalFindingBrief = {
  metric_type: string
  label: string
  status: string
  value_display: string
  trend_label: string
  priority: string
}

// ---------------------------------------------------------------------------
// ai-summary
// ---------------------------------------------------------------------------

export interface ClinicalSummaryOut {
  as_of: string
  conditions: string[]
  allergies: string[]
  medications: MedicationBrief[]
  abnormal_findings: AbnormalFindingBrief[]
  notable_changes: string[]
  sources: SourceRef[]
  confidence: ConfidenceLevel
  disclaimer: string
}

// ---------------------------------------------------------------------------
// ai-analysis
// ---------------------------------------------------------------------------

export interface RiskFlag {
  level: RiskLevel
  label_vi: string
  findings: string[]
  missing_data: string[]
  sources: SourceRef[]
}

export interface ClinicalAnalysisOut {
  priority: RiskFlag
  key_issues: string[]
  contradictions_or_gaps: string[]
  differentials_to_exclude: string[]
  confidence: ConfidenceLevel
  disclaimer: string
}

// ---------------------------------------------------------------------------
// ai-questions
// ---------------------------------------------------------------------------

export type SuggestedQuestionGroup =
  | 'current_symptoms'
  | 'onset_timing'
  | 'severity_progression'
  | 'aggravating_relieving'
  | 'relevant_history'
  | 'medication_adherence'
  | 'warning_signs'
  | 'lifestyle'

export interface SuggestedQuestion {
  group: SuggestedQuestionGroup
  question_vi: string
  reason_vi: string
}

export interface ClinicalQuestionsOut {
  questions: SuggestedQuestion[]
  confidence: ConfidenceLevel
  disclaimer: string
}

// ---------------------------------------------------------------------------
// ai-advice
// ---------------------------------------------------------------------------

export type AdviceCategory =
  | 'explain_patient'
  | 'home_monitoring'
  | 'when_to_visit'
  | 'suggested_tests'

export interface AdviceItem {
  category: AdviceCategory
  text_vi: string
}

export interface ClinicalAdviceOut {
  items: AdviceItem[]
  confidence: ConfidenceLevel
  disclaimer: string
}

// ---------------------------------------------------------------------------
// Feature flag + scope
// ---------------------------------------------------------------------------

/**
 * Feature flag. Defaults to `false` (disabled). Set
 * `NEXT_PUBLIC_CLINICAL_COPILOT=true` to enable the live call path.
 */
export const CLINICAL_COPILOT_ENABLED = process.env.NEXT_PUBLIC_CLINICAL_COPILOT === 'true'

/** Standing disclaimer shown on every copilot surface. */
export const CLINICAL_COPILOT_DISCLAIMER =
  'Gợi ý AI chỉ mang tính hỗ trợ. Bác sĩ chịu trách nhiệm quyết định chuyên môn.'

/** Scope that pins every copilot request to one patient (and optionally a consultation). */
export interface ClinicalCopilotScope {
  /** PatientProfile.id — the ONLY patient the copilot may reason about. */
  patientId: string
  /** Optional consultation context when invoked from a consultation. */
  consultationId?: string
}

// ---------------------------------------------------------------------------
// Typed functions
// ---------------------------------------------------------------------------

/**
 * Fetch the structured clinical summary for a patient.
 *
 * - When disabled (default): resolves `null` with NO network call.
 * - When enabled: POSTs to `/doctor/patients/{patientId}/ai-summary`. Real
 *   `ApiError`s propagate — callers must render an error state with retry.
 */
export async function getClinicalSummary(
  scope: ClinicalCopilotScope
): Promise<ClinicalSummaryOut | null> {
  if (!CLINICAL_COPILOT_ENABLED) return null
  return api.post<ClinicalSummaryOut>(`/doctor/patients/${scope.patientId}/ai-summary`, {
    consultation_id: scope.consultationId,
  })
}

/**
 * Fetch the structured risk analysis for a patient.
 *
 * - When disabled (default): resolves `null` with NO network call.
 * - When enabled: POSTs to `/doctor/patients/{patientId}/ai-analysis`.
 */
export async function getClinicalAnalysis(
  scope: ClinicalCopilotScope
): Promise<ClinicalAnalysisOut | null> {
  if (!CLINICAL_COPILOT_ENABLED) return null
  return api.post<ClinicalAnalysisOut>(`/doctor/patients/${scope.patientId}/ai-analysis`, {
    consultation_id: scope.consultationId,
  })
}

/**
 * Fetch suggested history-taking questions for a patient (optionally scoped
 * to a consultation).
 *
 * - When disabled (default): resolves `null` with NO network call.
 * - When enabled: POSTs to `/doctor/patients/{patientId}/ai-questions`.
 */
export async function getClinicalQuestions(
  scope: ClinicalCopilotScope
): Promise<ClinicalQuestionsOut | null> {
  if (!CLINICAL_COPILOT_ENABLED) return null
  return api.post<ClinicalQuestionsOut>(`/doctor/patients/${scope.patientId}/ai-questions`, {
    consultation_id: scope.consultationId,
  })
}

/**
 * Fetch a suggested counseling direction for a patient (optionally scoped
 * to a consultation).
 *
 * - When disabled (default): resolves `null` with NO network call.
 * - When enabled: POSTs to `/doctor/patients/{patientId}/ai-advice`.
 */
export async function getClinicalAdvice(
  scope: ClinicalCopilotScope
): Promise<ClinicalAdviceOut | null> {
  if (!CLINICAL_COPILOT_ENABLED) return null
  return api.post<ClinicalAdviceOut>(`/doctor/patients/${scope.patientId}/ai-advice`, {
    consultation_id: scope.consultationId,
  })
}
